### PR TITLE
Add 0.39.0 upgrade routine

### PR DESCRIPTION
**Pull request by VineethReddy02**
Part of #577

### CODE REVIEW

File: pkg/collector/upgrade/v0_39_0.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+package upgrade
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (jpkrohling) [inline pkg/collector/upgrade/v0_39_0.go:23]: The import grouping looks odd here.

File: pkg/collector/upgrade/v0_39_0_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+// limitations under the License.
+
+package upgrade_test
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (jpkrohling) [inline pkg/collector/upgrade/v0_39_0_test.go:22]: Import order looks odd here as well.

File: pkg/collector/upgrade/v0_39_0_test.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	assert.NoError(t, err)
+
+	assert.Equal(t, `processors:
+  memory_limiter: null
+  memory_limiter/with-settings:
+    check_interval: 5s
+    limit_mib: 4000
+    spike_limit_mib: 500
+receivers:
+  apache: null
```
Comment (jpkrohling) [inline pkg/collector/upgrade/v0_39_0_test.go:82]: Why the null here? Perhaps we should instead have an empty object?

**Review by jpkrohling — COMMENTED** [review]
Other than the import order, I'm just concerned that a `null` object might break the collector. We should probably have an empty object instead of null (or just nothing, like the input).

File: pkg/collector/upgrade/v0_39_0.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+func updateConfig(otelcol *v1alpha1.OpenTelemetryCollector, cfg map[interface{}]interface{}) (*v1alpha1.OpenTelemetryCollector, error) {
+	res, err := yaml.Marshal(cfg)
+	if err != nil {
+		return otelcol, fmt.Errorf("couldn't upgrade to v0.39.0, failed to marshall back configuration: %w", err)
+	}
+
+	// Note: This is a hack to drop null occurrences from our config parsing above in upgrade routine
+	otelcol.Spec.Config = strings.ReplaceAll(string(res), "null", "")
```
Comment (jpkrohling) [inline pkg/collector/upgrade/v0_39_0.go:117]: I wonder if this is going to have bad side effects. Like for someone posting traces to `mynullservice.com` :-)

Comment (VineethReddy02) [reply pkg/collector/upgrade/v0_39_0.go:117]: Maybe we should do `   null   `? with trailing spaces...

File: pkg/collector/upgrade/v0_24_0_test.go [hunk 5]
```diff
@@ -57,7 +57,7 @@ func TestHealthCheckEndpointMigration(t *testing.T) {
 
 	// verify
 	assert.Equal(t, `extensions:
-  health_check: null
```
Comment (VineethReddy02) [inline pkg/collector/upgrade/v0_24_0_test.go:60]: @jpkrohling the issue of `null` existed in the past too. The other contributor ignored it by adding `null` during the validation as the parser injects null to empty value.

**Review by jpkrohling — APPROVED** [review]

File: pkg/collector/upgrade/v0_24_0_test.go (hunk 5 above)
Comment (jpkrohling) [reply pkg/collector/upgrade/v0_24_0_test.go:60]: Can you please create an issue to track this?

**Pull request merged**